rust/agents/relayer/src/msg/metadata/aggregation.rs

```diff
@@ -0,0 +1,238 @@
+        let ism = self.build_aggregation_ism(ism_address).await.context(CTX)?;
+        let (modules, threshold) = ism.modules_and_threshold(message).await.context(CTX)?;
+        let threshold = threshold as usize;
+
+        let sub_isms: Vec<_> = join_all(modules.iter().map(|ism_address| {
+            try_join(
+                self.base.build(*ism_address, message),
+                self.base.build_ism(*ism_address),
+            )
+        }))
+        .await
+        .into_iter()
+        .enumerate()
+        .filter_map(|(i, r)| match r {
+            Ok((Some(meta), ism)) => Some((SubIsmMetadata::new(i, meta), ism)),
+            _ => None,
+        })
+        .collect();
+
+        Self::cheapest_valid_metas(sub_isms, message, threshold)
+            .await
+            .map(|maybe_proofs| {
+                maybe_proofs.map(|mut proofs| Self::format_metadata(&mut proofs, modules.len()))
+            })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use ethers::utils::hex::FromHex;
+
+    use super::*;
+
+    #[test]
+    fn test_format_n_of_n_metadata_works_correctly() {
+        let mut metadatas = vec![
+            SubIsmMetadata::new(
+                0,
+                Vec::from_hex("290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563")
+                    .unwrap(),
+            ),
+            SubIsmMetadata::new(
+                1,
+                Vec::from_hex("510e4e770828ddbf7f7b00ab00a9f6adaf81c0dc9cc85f1f8249c256942d61d9")
+                    .unwrap(),
+            ),
+            SubIsmMetadata::new(
+                2,
+                Vec::from_hex("356e5a2cc1eba076e650ac7473fccc37952b46bc2e419a200cec0c451dce2336")
+                    .unwrap(),
+            ),
+        ];
+        let expected = Vec::from_hex("000000180000003800000038000000580000005800000078290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563510e4e770828ddbf7f7b00ab00a9f6adaf81c0dc9cc85f1f8249c256942d61d9356e5a2cc1eba076e650ac7473fccc37952b46bc2e419a200cec0c451dce2336").unwrap();
+        assert_eq!(
+            AggregationIsmMetadataBuilder::format_metadata(&mut metadatas, 3),
+            expected
+        );
+    }
+
+    #[test]
+    fn test_format_n_of_m_metadata_works_correctly() {
+        let mut metadatas = vec![
+            SubIsmMetadata::new(
+                0,
+                Vec::from_hex("290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563")
+                    .unwrap(),
+            ),
+            SubIsmMetadata::new(
+                1,
+                Vec::from_hex("510e4e770828ddbf7f7b00ab00a9f6adaf81c0dc9cc85f1f8249c256942d61d9")
+                    .unwrap(),
+            ),
+            SubIsmMetadata::new(
+                2,
+                Vec::from_hex("356e5a2cc1eba076e650ac7473fccc37952b46bc2e419a200cec0c451dce2336")
+                    .unwrap(),
+            ),
+            SubIsmMetadata::new(
+                4,
```

Comment:
is this intended?

Reply:
It's intended yep, there's a total of 5 ISMs but only 4 are required, so the one with index 3 is not included. I added a comment.